### ISSUE TITLE
Compiler panic when detecting cycle

### ISSUE THREAD

**Issue by dvdplm** [rust-lang/rust#51711]
Traits with cycles between them cause an ICE, with erratic behaviour wrt to whitespace (!).

Expected: a compiler error about cyclic trait dependencies.

Compiler version:
```
$ rustc --version --verbose
rustc 1.28.0-nightly (f28c7aef7 2018-06-19)
binary: rustc
commit-hash: f28c7aef7fbff1909c2d0257186cd7a5f0c6aa4b
commit-date: 2018-06-19
host: x86_64-apple-darwin
release: 1.28.0-nightly
LLVM version: 6.0
```

This ICEs consistently:

```
pub trait Hasher {}

pub trait HashDB<H: Hasher>: AsHashDB<H> {}
pub trait AsHashDB<H: Hasher>: std::convert::AsRef<HashDB<H>> {}
```

This does not:
```
pub trait Hasher {}
pub trait HashDB<H: Hasher>: AsHashDB<H> {}
pub trait AsHashDB<H: Hasher>: std::convert::AsRef<HashDB<H>> {}
```

Interestingly this seems to be a MacOS-specific bug.

Backtrace:
```
thread 'main' panicked at 'did not find a cycle', librustc/ty/query/job.rs:143:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
   6: std::panicking::begin_panic
   7: rustc::ty::query::job::QueryJob::await
   8: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
   9: <rustc::traits::util::SupertraitDefIds<'cx, 'gcx, 'tcx> as core::iter::iterator::Iterator>::next
  10: rustc::traits::object_safety::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::astconv_object_safety_violations
  11: <rustc_typeck::astconv::AstConv<'gcx, 'tcx> + 'o>::conv_object_ty_poly_trait_ref
  12: <rustc_typeck::astconv::AstConv<'gcx, 'tcx> + 'o>::ast_ty_to_ty
  13: rustc::ty::subst::<impl rustc::ty::Slice<rustc::ty::subst::Kind<'tcx>>>::fill_item
  14: rustc::ty::subst::<impl rustc::ty::Slice<rustc::ty::subst::Kind<'tcx>>>::for_item
  15: <rustc_typeck::astconv::AstConv<'gcx, 'tcx> + 'o>::create_substs_for_ast_path
  16: rustc::hir::PathSegment::with_parameters
  17: <rustc_typeck::astconv::AstConv<'gcx, 'tcx> + 'o>::create_substs_for_ast_trait_ref
  18: <rustc_typeck::astconv::AstConv<'gcx, 'tcx> + 'o>::instantiate_poly_trait_ref_inner
  19: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter
  20: rustc_typeck::collect::compute_bounds
  21: rustc_typeck::collect::super_predicates_of
  22: rustc::ty::query::__query_compute::super_predicates_of
  23: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::super_predicates_of<'tcx>>::compute
  24: rustc::ty::context::tls::with_context
  25: rustc::ty::context::tls::with_context
  26: rustc::dep_graph::graph::DepGraph::with_task_impl
  27: rustc::ty::context::tls::with_related_context
  28: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  29: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query
  30: rustc::ty::query::plumbing::force_from_dep_node
  31: rustc::dep_graph::graph::DepGraph::try_mark_green
  32: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::try_mark_green_and_read
  33: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  34: rustc::ty::query::TyCtxtAt::super_predicates_of
  35: <rustc_typeck::collect::CollectItemTypesVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item
  36: rustc::hir::Crate::visit_all_item_likes
  37: rustc::session::Session::track_errors
  38: rustc_typeck::check_crate
  39: rustc::ty::context::tls::enter_context
  40: <std::thread::local::LocalKey<T>>::with
  41: rustc::ty::context::TyCtxt::create_and_enter
  42: rustc_driver::driver::compile_input
  43: rustc_driver::run_compiler_with_pool
  44: <scoped_tls::ScopedKey<T>>::set
  45: syntax::with_globals
  46: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  47: __rust_maybe_catch_panic
  48: rustc_driver::run
  49: rustc_driver::main
  50: std::rt::lang_start::{{closure}}
  51: std::panicking::try::do_call
  52: __rust_maybe_catch_panic
  53: std::rt::lang_start_internal
  54: main
query stack during panic:
#0 [super_predicates_of] computing the supertraits of `AsHashDB`
end of query stack
```

**Comment by xfix** [comment]
The function that is ICE-ing appears to be annotated `#[cfg(not(parallel_queries))]`, which may be meaningful for reproducing the issue.

**Comment by dvdplm** [comment]
I can no longer repeat this with last night's build: `rustc 1.28.0-nightly (cbc4c8380 2018-06-22)` and even if I `rustup toolchain install nightly-2018-06-19` I don't get the same rust version. Instead I get

```
rustc 1.28.0-nightly (523097979 2018-06-18)
binary: rustc
commit-hash: 5230979794db209de492b3f7cc688020b72bc7c6
commit-date: 2018-06-18
host: x86_64-apple-darwin
release: 1.28.0-nightly
LLVM version: 6.0
```

…and this version does not exhibit the issue. Was the faulty version pulled perhaps?

**Comment by Mark-Simulacrum** [comment]
Generally the version you need to get from rustup is off by one on the date, IIRC by one forward -- so the nightly in this case would probably be nightly-2018-06-20.

**Comment by rylev** [comment]
I cannot reproduce this. Note that the commit-hash from the nightly version available through rustup is not the same as the one in the original issue post. 
```
rustc --version --verbose
rustc 1.28.0-nightly (f28c7aef7 2018-06-19)
binary: rustc
commit-hash: f28c7aef7fbff1909c2d0257186cd7a5f0c6aa4b
commit-date: 2018-06-19
host: x86_64-apple-darwin
release: 1.28.0-nightly
LLVM version: 6.0
```

```
cat src/lib.rs
pub trait Hasher {}

pub trait HashDB<H: Hasher>: AsHashDB<H> {}
pub trait AsHashDB<H: Hasher>: std::convert::AsRef<HashDB<H>> {}
```

This produces the correct error:
```
cargo build
   Compiling deleteme v0.1.0 (file:///Users/ryanlevick/deleteme)
error[E0391]: cycle detected when computing the supertraits of `HashDB`
 --> src/lib.rs:3:1
  |
3 | pub trait HashDB<H: Hasher>: AsHashDB<H> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
note: ...which requires computing the supertraits of `AsHashDB`...
 --> src/lib.rs:4:1
  |
4 | pub trait AsHashDB<H: Hasher>: std::convert::AsRef<HashDB<H>> {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = note: ...which again requires computing the supertraits of `HashDB`, completing the cycle

error: aborting due to previous error

For more information about this error, try `rustc --explain E0391`.
error: Could not compile `deleteme`.

To learn more, run the command again with --verbose
```

**Comment by dvdplm** [comment]
Let's close this. :)
